Create /devtools to list useful development URLs

cory2067 opened the pull request:
Added a link to this page on `/devlogin`.
Additionally, updated the tools themselves to output the URL of the output.

The page may overlap with the header, but this will be resolved upon merging #5376.

smarx commented:
*Automated message from Dropbox CLA bot*

@cory2067, it looks like you've already signed the Dropbox CLA. Thanks!

YagoGG commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,40 @@
+<div id="devtools-page">
+    <h2>Useful development URLs</h2>
+    <p>
+        Below is a list of useful URLs that provide information useful in development.
```
Not sure if this is too obvious, but maybe it could be worth mentioning that the development server must be running to be able to see these?

I don't know if it may be confusing for people e.g. running only `./tools/test-js-with-node --coverage` and having problems when going directly to http://localhost:9991/node-coverage/index.html.

cory2067 replied:
Well, if you're viewing this page, then your dev server must be running :P
I guess it couldn't hurt to add.

YagoGG commented:
Hi @cory2067! Thanks for working on this.

I just tested the feature in my development environment and it looks great! Left a small inline comment.

Would you mind tagging the strings in the HTML template for translation? You can find some guidance on how to do it [here](http://zulip.readthedocs.io/en/latest/translating.html#backend-translations).

cory2067 commented:
Ok, done!

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+        </thead>
+        <tbody>
+            <tr>
+                <td><code>./tools/test-backend</code></td>
```
needs `--coverage` -- it doesn't do coverage by default.

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+            </tr>
+            <tr>
+                <td>None needed</td>
+                <td><a href="/test_emails">/test_emails</a></td>
```
I think this is /emails.

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+        Most of these require you to run a command to build the relevant content. This
+        table specifies which command to use for each page. Make sure your development server
+        is still running when you visit these!
+        {% endtrans %}
```
We generally don't tag for translation the development environment specific pages, just to save translations low-value translations.

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+        {% trans %}
+        Below is a list of useful URLs that provide information useful in development.
+        Most of these require you to run a command to build the relevant content. This
+        table specifies which command to use for each page. Make sure your development server
```
"which command to use to update the data served by each page (since several of these, like test coverage, require a special command to be run to generate the data)"

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+            <tr>
+                <th>Command</th>
+                <th>URL</th>
+            </tr>
```
I might swap the column order, and maybe also add a brief description for each?

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+    <h2>Useful development URLs</h2>
+    <p>
+        {% trans %}
+        Below is a list of useful URLs that provide information useful in development.
```
Maybe, "Below is a list of useful tools and data sets available only in the Zulip development environment that are often useful when contributing to Zulip.  "

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+    <p>
+        {% trans %}
+        Below is a list of useful URLs that provide information useful in development.
+        Most of these require you to run a command to build the relevant content. This
```
s/build/generate/

timabbott commented on docs/Makefile:
```diff
@@ -55,6 +55,7 @@ html:
 	$(SPHINXBUILD) -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
+	@echo "View pages at http://127.0.0.1:9991/docs/index.html"
```
One thing we may want to mention here is that the index doesn't update automatically for pages that aren't build rebuilt, and they might need to run `rm -rf docs/_build` to update that as well.

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,43 @@
+                <td><a href="/node-coverage/index.html">/node-coverage/index.html</a></td>
+            </tr>
+            <tr>
+                <td><code>cd docs &amp;&amp; make html</code></td>
```
I think we might want to replace this with a tiny `tools/build-docs` wrapper, just for similarity with the rest.

cory2067 commented:
Thanks for the notes. I believe I addressed all these, and added a `./tools/build-docs` which also deletes `_build_` if it exists before making the docs.

timabbott commented on tools/build-docs:
```diff
@@ -0,0 +1,7 @@
+rm -rf _build
+make html
+cd ..
+s
```
When writing shell scripts like this, you should never assume anything about the current working directory that the user is running it from, because it'll fail for those users.  See `tools/test-all` for an example: It does `cd "$(dirname "$0")"/..` at the top to address this.

Also, what's the `s`?

(As a follow-up project, it may be worth doing a quick audit from the shell scripts under `tools/` for this issue; `tools/test-migrations` at least has it as well)

timabbott commented on templates/zerver/dev_login.html:
```diff
@@ -36,6 +36,9 @@ <h4 class="login-page-header">{{ _('Click on a user to log in!') }}</h4>
             </div>
         </form>
+        <div id="devtools-wrapper">
+            <a id="devtools-link" href="/devtools">Zulip Developer Tools</a>
```
Stylistically, we usually capitalized just the first word in this sort of title.

timabbott commented on templates/zerver/dev_tools.html:
```diff
@@ -0,0 +1,49 @@
+                <td>View Zulip documentation</td>
+            </tr>
+            <tr>
+                <td><a href="/emails">/test_emails</a></td>
```
oh, you missed a reference to `/test_emails` here.

cory2067 commented:
updated

timabbott commented:
@cory2067 the final result of this branch is correct (modulo one issue), but you should be ordering your commits so that each one is correct at the time.  So, for example, `tools/build-docs` should have been rebased to the bottom of the branch so that 

This is partially about making life better when partially merging code, but it's also an important part of communicating clearly.

I dropped one commit (will post an explanation for why shortly) and merged this, after fixing that commit ordering issue.  Thanks @cory2067!  

One small follow-up: we should probably mention the /devtools/ in a few places in the developer documentation (maybe where we mention the tools it references, e.g. near coverage and emails), just to provide extra bread crumbs for it, since it seems like one could easily miss its existence if not paying enough attention.

timabbott commented on zerver/lib/test_helpers.py:
```diff
@@ -421,6 +421,7 @@ def find_pattern(pattern, prefixes):
 
         if full_suite:
             print('INFO: URL coverage report is in %s' % (fn,))
+            print('INFO: View report at http://127.0.0.1:9991/coverage/index.html')
```
This section of code in `test_helpers.py` is for the "URL coverage" report, which is separate from the "test coverage" report, and unrelated to that URL.
